Fix #6137: add `-Wincomplete-record-updates`

andreasabel opened the pull request:
Aiming to fix #6137.

Refactoring steps:
1. Break `Defn` into a tagged union or records.  Needs the `COMPLETE` pragma which requires GHC >= 8.2.
2. Refine `emptyFunction` to `emptyFunctionData`.
3. ... (see commits)

andreasabel committed:
Refactor #6137: break Defn into a tagged union of records

nad commented:
Did you use the `COMPLETE` pragma?

andreasabel committed:
Refactor #6137: `emptyFunctionData :: FunctionData`

Use `emptyFunctionData` rather than `emptyFunction` when updating fields
of the empty function.

andreasabel commented:
> Did you use the `COMPLETE` pragma?

Thanks this fixed my problem which was:

_Unfortunately, the coverage checker of GHC does not seem smart enough to certify coverage with the pattern synonyms I defined._
```
[175 of 414] Compiling Agda.TypeChecking.Monad.Base ( src/full/Agda/TypeChecking/Monad/Base.hs, nothing ) [Source file changed]

src/full/Agda/TypeChecking/Monad/Base.hs:2590:3: error: [-Wincomplete-patterns, -Werror=incomplete-patterns]
    Pattern match(es) are non-exhaustive
    In an equation for ‘pretty’:
        Patterns of type ‘Defn’ not matched:
            AxiomDefn _
            DataOrRecSigDefn _
            FunctionDefn _
            DatatypeDefn _
            ...
     |
2590 |   pretty Axiom{} = "Axiom"
     |   ^^^^^^^^^^^^^^^^^^^^^^^^...
```

nad commented:
In another currently open pull request (#6099), in which I changed a key data type (`Term`) and used a pattern synonym to avoid breakage, I turned off coverage checking for GHC 8.8.4 and earlier to avoid problems. It seems as if the coverage checker in GHC 8.10.7 is better.

andreasabel committed:
Refactor #6137: `emptyFunction` in `Data.hs`; `instance Pretty Defn`

andreasabel commented:
Thanks.  Do you mean to say that a `COMPLETE` pragma isn't necessary from GHC 8.10.7?

We are going to drop GHC 8.0 fairly soon, which makes the `COMPLETE` pragma available.
Alternatively, we can just switch off coverage checking for GHC 8.0.

andreasabel committed:
Refactor #6137: use DatatypeData

andreasabel committed:
Refactor #6137: remaining refactorings concerning Defn

andreasabel committed:
Refactor #6137: turn EqualityView into tagged sum of records

andreasabel committed:
Refactor #6137: complete-record-updates involving `C.NoName`

andreasabel committed:
Fix #6137: add `-Wincomplete-record-updates`; limit coverage checking to GHC 9.x

The coverage checking for GHC 8.x is too weak to pass this refactoring (#6142).

nad commented:
> Do you mean to say that a `COMPLETE` pragma isn't necessary from GHC 8.10.7?

No. See my comment in [Agda.cabal](https://github.com/agda/agda/pull/6099/files#diff-5244399240e85b97221fecc4a10cfbdc6fa2215d73fa99c127a37e07485ece22).

andreasabel commented:
> ... to avoid breakage, I turned off coverage checking for GHC 8.8.4 and earlier to avoid problems. It seems as if the coverage checker in GHC 8.10.7 is better.

Looks like the coverage checker even improved in GHC 9.0.2: https://github.com/agda/agda/actions/runs/3151460777

Thus, we can turn it off for GHC 8.x and declare GHC 9.x as sole compilers for Agda development.

jespercockx commented on src/full/Agda/TypeChecking/Monad/Base.hs:
```diff
@@ -4963,3 +5329,5 @@ instance NFData UnificationFailure
 instance NFData TypeError
 instance NFData LHSOrPatSyn
+
+-- -}
```
This could be cleaned up, if I'm allowed to nitpick.

andreasabel replied:
I think this should be the official EOF marker.  ;-)

andreasabel replied:
Fixed that.

nad commented on src/full/Agda/TypeChecking/Monad/Base.hs:
```diff
@@ -2182,152 +2182,423 @@ defaultAxiom = Axiom False
+    funTerminates
+    funExtLam
+    funWith
+  )
```
It might be a bit easier to get an overview of the code if the pattern synonyms are moved out of the way.

andreasabel replied:
I think `Definition` and all types leading up to it should go into a new module `Agda.Typechecking.Signature.Types`, and then we have a module `Agda.TypeChecking.Signature.Tools` for all that small editors, pattern synonyms etc.
